### PR TITLE
Comprehensive, Native Windows Support

### PR DESCRIPTION
**Pull request by Alexhuszagh**
Here is a tentative pull request for comprehensive, native Windows support for the enhancement in #519. It passes all the current leveldb unit tests, and compiles on Visual Studio 2015 and 2017, and MinGW with GCC 5.4+. A list of the possibly notable differences is described below:

- Use `CRITICAL_SECTION` rather than WinAPI mutexes (reused from the Boost version by Gabor Cselle).
- Uses a semaphore-based condition variable for Windows XP support (reused from the Boost version by Gabor Cselle).
- Accepts UTF-8 paths (internally converts to UTF-16 for Windows Unicode path API).
- Supports sequential, random-access, and memory-mapped files.
- Does not support extended-length paths (prefixed with `\\?\`), due the its [limitations and cost](https://github.com/google/leveldb/issues/519#issuecomment-342437294).
- `WindowsEnv::GetChildren` ignores `.` and `..`, but not other dot files, by default.

A few caveats:
- Relies on `std::thread` and `std::chrono`, and therefore requires C++11 support.
- `c_test` was disabled on Windows due to its reliance on `tmpnam`, since MinGW has a malformed `tmpnam` implementation.
- `db_test` disables the check for `.` and `..` on Windows only after creating a directory for `GetChildren`, since the WindowsEnv currently filters these items out.
- The MakeFile system does not work (this is a general issue with [pwnnall/cmake](https://github.com/pwnall/leveldb/tree/cmake)). I can provide patches that would maintain compatibility with the MakeFile system, but have not yet done so.

You can see the successful builds on Linux, macOS, and Windows on the [Travis CI](https://travis-ci.org/Alexhuszagh/leveldb/builds/299314513) and [AppVeyor](https://ci.appveyor.com/project/Alexhuszagh/leveldb/build/22) pages for my leveldb fork.

### CODE REVIEW

**Comment by Alexhuszagh** [comment]
Just wondering if you will have time to review this pull request at any time in the near future, @pwnall. Sorry for the notification.

**Comment by ghemawat** [comment]
Quick question: there seem to be a ton of files that are showing up in this
PR, but have no changes.  Is there some easy way to fix that up?

I am going to take a quick look right now at the header file changes to see
if there are any significant/wide-spread issues with the PR.  And we will
try to look at the rest of the PR soon.

File: port/port_windows.h [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+  void Lock();
+  void Unlock();
+  void AssertHeld();
+
+ private:
+  friend class CondVar;
+  // critical sections are more efficient than mutexes
+  // but they are not recursive and can only be used to synchronize threads within the same process
+  // we use opaque void * to avoid including windows.h in port_win.h
+  void * cs_;
```
Comment (ghemawat) [inline port/port_windows.h]: Any particular reason to not use C++11's std::mutex?

I don't undestand the "critical section are more efficient than mutexes" comment.  Which mutex implementation is that talking about, and in which scenarios are they less efficient?

File: port/port_windows.h [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+  void Signal();
+  void SignalAll();
+ private:
+  Mutex* mu_;
+
+  Mutex wait_mtx_;
+  long waiting_;
+
+  void * sem1_;
+  void * sem2_;
```
Comment (ghemawat) [inline port/port_windows.h]: Is C++11's std::condition_variable unsuitable for some reason?

File: port/port_windows.h [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+
+
+};
+
+// Storage for a lock-free pointer
+class AtomicPointer {
+ private:
+  void * rep_;
```
Comment (ghemawat) [inline port/port_windows.h]: IIRC, leveldb is planning to switch to C++11.  Then, we can use std::atomic<void*> instead and not need AtomicPointer.

Can we rely on C++11 for the windows port instead and then just arrange to use the std:atomic<> based AtomicPointer defined in port/atomic_pointer.h?

File: port/port_windows.cc [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+}
+
+void CondVar::Signal() {
+  wait_mtx_.Lock();
+  if (waiting_ > 0) {
+    --waiting_;
+
+    // finalize handshake
+    ::ReleaseSemaphore(sem1_, 1, NULL);
+    ::WaitForSingleObject(sem2_, INFINITE);
```
Comment (ghemawat) [inline port/port_windows.cc]: This is tricky code.  I would be much happier if we can use some existing library.  std::condition_variable seems the most appropriate to me.

File: port/port_windows.cc [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+  }
+  wait_mtx_.Unlock();
+}
+
+void CondVar::SignalAll() {
+  wait_mtx_.Lock();
+  for(long i = 0; i < waiting_; ++i) {
+    ::ReleaseSemaphore(sem1_, 1, NULL);
+    while(waiting_ > 0) {
```
Comment (ghemawat) [inline port/port_windows.cc]: I don't understand this code.  Suppose waiting_ is initially 2.  The sequence of operations will be:

ReleaseSemaphore(sem1_, 1, NULL)
WaitForSingleOnbject(sem2_, INF)
WaitForSingleOnbject(sem2_, INF)

Don't we instead want paired Release/Wait calls?

Even if that is fixed, I am not very confident in my understanding/trust of this code.  A ready-made implementation we can call would be better.

File: port/port_windows.cc [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+  wait_mtx_.Unlock();
+}
+
+AtomicPointer::AtomicPointer(void* v) {
+  Release_Store(v);
+}
+
+void* AtomicPointer::Acquire_Load() const {
+  void * p = nullptr;
+  InterlockedExchangePointer(&p, rep_);
```
Comment (ghemawat) [inline port/port_windows.cc]: Seems unnecessarily expensive, especially on x86-es where an acquire load can be much faster than an compare-and-swap.

File: util/windows_logger.h [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+#include <windows.h>
+#include "leveldb/env.h"
+
+#ifdef DeleteFile   // Win32 defines the DeleteFile macro
+#undef DeleteFile
+#endif
+
+namespace leveldb {
+
+class WindowsLogger : public Logger {
```
Comment (ghemawat) [inline util/windows_logger.h]: This looks pretty similar to posix_logger.  I wonder if we can restructure things to share that code.  That code already has removed the gettid dependency by taking a function pointer.  We could do a similar thing for getting the current local time into a string.

File: port/port_windows.cc (hunk 5 above)
Comment (Alexhuszagh) [reply port/port_windows.cc]: I'm going to simply comment for all this code: this was repurposed from the Windows branch [already in leveldb](https://github.com/google/leveldb/blob/windows/port/port_win.cc#L92) and exists for the purpose of avoiding std::mutex, since the `CONDITION_VARIABLE` didn't exist until Windows Vista. This is likely a non-issue. See my comments on the `CRITICAL_SECTION` for why I use this rather than `std::condition_variable`, although after your comments, I might swap back.

File: port/port_windows.h (hunk 1 above)
Comment (Alexhuszagh) [reply port/port_windows.h]: Various [programmers](https://stoyannk.wordpress.com/2016/04/30/msvc-mutex-is-slower-than-you-might-expect/) have dug down and found `std::mutex` can be significantly slower than a manual `SRW_LOCK` or `CRITICAL_SECTION`, even under ideal situations due to MSVC's implementation. MSVC 2015, the first with true C++11 support, mostly mitigates this with a better implementation that uses `SRW_LOCK` under the hood.

> TLDR; The 2015 MSVC C++ runtime’s std::mutex (and potentially other) implementation is significantly slower than the equivalent code written by hand. The reason is that the runtime is built with a Windows feature called “Control Flow Guard” and uses function pointers.... It turned out that when the CRT calls __crtAcquireSRWLockExclusive (and all other SRWLOCK methods) they don’t go directly in the Windows Kernel32.dll! A lot of checks and code is executed between entering the method and actually arriving in AcquireSRWLockExclusive, which is where we want to go. The reason is a Windows feature called “Control Flow Guard“. Essentially this is a security feature that instructs the compiler to check every jump through a function pointer and validate if it is a valid target. This is possible because on modern Windows, all function addresses are annotated and known to the loader. The feature will prevent jumping “in the middle” of functions and makes hacking more difficult.

In short, this may well be worth the cost of a slight runtime cost, however, I avoided it at first to avoid any performance penalties. I also ran a few benchmarks which were generally consistent with the conclusions of the author (using MSVC 2015, on Windows 10). I would gladly reimplement it as `std::mutex` and `std::condition_variable` if that slight runtime cost is worth decreased code complexity.

File: port/port_windows.h (hunk 3 above)
Comment (Alexhuszagh) [reply port/port_windows.h]: I'll switch this, thank you.

File: util/windows_logger.h (hunk 7 above)
Comment (Alexhuszagh) [reply util/windows_logger.h]: I'll work on this as well, and edit my commit history to reflect this. Thanks.

**Comment by solvingj** [comment]
It looks like this is the most comprehensive of 3 PR's relating to windows support.  Is this PR still under consideration?

**Comment by Alexhuszagh** [comment]
@solvingj I recently got backlogged with work, and will be addressing all of Sanjay's suggestions in short order. Sorry about the delay.

**Comment by Alexhuszagh** [comment]
@ghemawat I have finished everything besides sharing code between the Windows and POSIX logger (sorry, I finally got free time today), but I had a small question:

Since I am now using only C++11 features for `Mutex`, `CondVar`, and `AtomicPointer` (it avoids any refactoring to use the existing code, but only as shallow wrappers around C++11 features), I was wondering if I should remove all the platform-specific port files and make all implementations use the C++11 features (likely as `port.h` and `port.cc`). Thanks, and sorry for the delay.

File: port/port_windows.cc (hunk 4 above)
Comment (Alexhuszagh) [reply port/port_windows.cc]: @ghemawat This has been removed. See the new code. I've used a centralized `port_cpp11` for this, and the new code is:

```cpp
std::unique_lock<std::mutex> lock(mu_->mu_, std::defer_lock);
cv_.wait(lock);
```

The rationale is the lock is already locked, and condition variable needs a unique_lock. Fortunately, `defer_lock` allows us to avoid unlocking and relocking the mutex. There is still an issue here though, in that the mutex is unlocked after waiting, which was not true previously (danger zone).

I am contemplating changing the above code to:

```cpp
{
    std::unique_lock<std::mutex> lock(mu_->mu_, std::defer_lock);
    cv_.wait(lock);
}
mu_->mu_->Lock();
```

This adds a redundant lock call, but reinstates correct behavior. Some comment would be great here.

File: port/port_windows.cc (hunk 5 above)
Comment (Alexhuszagh) [reply port/port_windows.cc]: @ghemawat I did some benchmarks and found no noticeable performance differences in LevelDB, at least for Wine. Everything uses the C++11 ports now in my version.

File: port/port_windows.cc (hunk 6 above)
Comment (Alexhuszagh) [reply port/port_windows.cc]: @ghemawat I've changed this to use `std::atomic::load`, which should be much more efficient.

File: util/windows_logger.h (hunk 7 above)
Comment (Alexhuszagh) [reply util/windows_logger.h]: @ghemawat I've refactored them, changed `PosixLogger` to be named `FileLogger`, and added a method called `AppendDateTo` which takes an iterator range for the first and last character in the buffer. This function differs on Windows and POSIX, but allows the rest of the code to be shared between POSIX and Windows.

**Windows**

```cpp
char* FileLogger::AppendDateTo(char* first, char* last)
{
  const uint64_t thread_id = (*gettid_)();

  SYSTEMTIME st;
  GetLocalTime(&st);
  first += snprintf(first, last - first,
                "%04d/%02d/%02d-%02d:%02d:%02d.%03d %llx ",
                st.wYear,
                st.wMonth,
                st.wDay,
                st.wHour,
                st.wMinute,
                st.wSecond,
                st.wMilliseconds,  // No microsecond precision, use milliseconds
                static_cast<long long unsigned int>(thread_id));

  return first;
}
```

**POSIX**

```cpp
char* FileLogger::AppendDateTo(char* first, char* last)
{
  const uint64_t thread_id = (*gettid_)();

  struct timeval now_tv;
  gettimeofday(&now_tv, NULL);
  const time_t seconds = now_tv.tv_sec;
  struct tm t;
  localtime_r(&seconds, &t);

  first += snprintf(first, last - first,
                "%04d/%02d/%02d-%02d:%02d:%02d.%06d %llx ",
                t.tm_year + 1900,
                t.tm_mon + 1,
                t.tm_mday,
                t.tm_hour,
                t.tm_min,
                t.tm_sec,
                static_cast<int>(now_tv.tv_usec),
                static_cast<long long unsigned int>(thread_id));

  return first;
}
```

**Comment by Alexhuszagh** [comment]
Sorry for the massive delay, most of these concerns should finally be addressed. I recently started a new job, and got bogged down with work, so personal projects of mine were temporarily stalled.

As for the CLA bot, I believe it's concerned about @pwnall's CLA, which should be a non-issue, since I believe he is a Google employee? 

Again, thanks. Any comments would be wonderful, when you have time.

**Comment by ghemawat** [comment]
I apologize, but due to some other deadlines and events, I won't be able to
get to this until late next week.

**Comment by ghemawat** [comment]
On Fri, Jan 19, 2018 at 1:36 AM Alexander Huszagh <notifications@github.com>
wrote:

> @ghemawat <https://github.com/ghemawat> I have finished everything
> besides sharing code between the Windows and POSIX logger (sorry, I finally
> got free time today), but I had a small question:
>
> Since I am now using only C++11 features for Mutex, CondVar, and
> AtomicPointer (it avoids any refactoring to use the existing code, but
> only as shallow wrappers around C++11 features), I was wondering if I
> should remove all the platform-specific port files and make all
> implementations use the C++11 features (likely as port.h and port.cc).
> Thanks, and sorry for the delay.
>

Even if C++11 support is available, sometimes there are reasons to use
different implementations of Mutex/CondVar which provide extra features
(like automatic lock contention profiling, etc.)  So we would prefer to
keep the the portability shim for Mutex/CondVar.

I think AtomicPointer is less useful as a portability layer, and we will
probably want to remove it once we have been requiring C++11 for a while.
But I think there isn't much reason to mix that change into this pull
request, so I suggest you leave the portability layer as it is for now.

File: util/env_windows.cc [hunk 8]
```diff
@@ -0,0 +1,920 @@
...
+  buffer[size] = L'\0';
+  MultiByteToWideChar(CP_UTF8, 0, &narrow[0], narrow.size(), &buffer[0], size);
+  return buffer;
+}
+
+// Helper class to limit resource usage to avoid exhaustion.
+// Currently used to limit read-only file descriptors and mmap file usage
+// so that we do not end up running out of file descriptors, virtual memory,
+// or running into kernel performance problems for very large databases.
+class Limiter {
```
Comment (ghemawat) [inline util/env_windows.cc]: We could consider moving Limiter into a helper module of its own (util/limiter.{h,cc}).  Are there any differences from the Posix version?

File: util/env_windows.cc [hunk 9]
```diff
@@ -0,0 +1,920 @@
...
+        OPEN_EXISTING,
+        FILE_FLAG_RANDOM_ACCESS,
+        NULL
+      );
+      if (handle == INVALID_HANDLE_VALUE) {
+        return WindowsError(filename_, GetLastError());
+      }
+    }
+
+    pread_mutex.Lock();
```
Comment (ghemawat) [inline util/env_windows.cc]: If OVERLAPPED ReadFile does not work, consider just making the pread_mutex a field of WindowsRandomAccessFile so that every different file object gets its own mutex.  I think you are only trying to protect the file offset.

File: util/env_windows.cc [hunk 10]
```diff
@@ -0,0 +1,920 @@
...
+    if (temporary_handle_) {
+      // Close the temporary file descriptor opened earlier.
+      CloseHandle(handle);
+    }
+    return s;
+  }
+};
+
+
+class WindowsMmapReadableFile: public RandomAccessFile
```
Comment (ghemawat) [inline util/env_windows.cc]: I think we should leave out the mmap support in this initial PR.  Consider a separate PR for that later on backed by real performance data as to how much mmap based reading helps.

File: util/env_windows.cc [hunk 11]
```diff
@@ -0,0 +1,920 @@
...
+  }
+
+  virtual Status Flush() {
+    if (!::FlushFileBuffers(handle_)) {
+      return WindowsError(filename_, GetLastError());
+    }
+    return Status::OK();
+  }
+
+  Status SyncDirIfManifest() {
```
Comment (ghemawat) [inline util/env_windows.cc]: Since SyncDirIfManifest() is empty, just delete it and move the comment into Sync.

File: util/env_windows.cc [hunk 12]
```diff
@@ -0,0 +1,920 @@
...
+
+
+class WindowsFileLock : public FileLock {
+ public:
+  HANDLE handle_;
+  std::string name_;
+};
+
+// Set of locked files.  We keep a separate set instead of just
+// relying on fcntrl(F_SETLK) since fcntl(F_SETLK) does not provide
```
Comment (ghemawat) [inline util/env_windows.cc]: fcntl comments seem stale.

Also, is this lock table really needed on Windows?  I am not sure if LockFile/UnlockFile have the same problem as fcntl(F_SETLK) does.

File: util/env_windows.cc [hunk 13]
```diff
@@ -0,0 +1,920 @@
...
+    locked_files_.erase(fname);
+  }
+};
+
+
+class WindowsEnv : public Env
+{
+ public:
+  WindowsEnv();
+  virtual ~WindowsEnv() {
```
Comment (ghemawat) [inline util/env_windows.cc]: This method is not supposed to be called.  So you can delete all code from here to the
   char msg[] = ...

File: util/env_windows.cc [hunk 14]
```diff
@@ -0,0 +1,920 @@
...
+    result->clear();
+
+    WIN32_FIND_DATAW data;
+    HANDLE handle = FindFirstFileW(Utf8ToUtf16(dir + "/*").data(), &data);
+    if (handle == INVALID_HANDLE_VALUE) {
+      return WindowsError(dir, GetLastError());
+    }
+    while (FindNextFileW(handle, &data) != 0) {
+      std::vector<char> v = Utf16ToUtf8(data.cFileName);
+      if (!(strcmp(v.data(), ".") == 0 || strcmp(v.data(), "..") == 0)) {
```
Comment (ghemawat) [inline util/env_windows.cc]: Any particular reason to have this extra check here that drops "." and ".." and then have to deal with the difference from Posix in one of the tests you updated?

File: util/env_windows.cc [hunk 15]
```diff
@@ -0,0 +1,920 @@
...
+
+  virtual void StartThread(void (*function)(void* arg), void* arg);
+
+  virtual Status GetTestDirectory(std::string* result) {
+    const char* env = getenv("TEST_TMPDIR");
+    if (env && env[0] != '\0') {
+      *result = env;
+    } else {
+      char buf[256];
+      int size = snprintf(buf, sizeof(buf), "%s/leveldbtest-%d", getenv("TMP"), int(GetCurrentProcessId()));
```
Comment (ghemawat) [inline util/env_windows.cc]: what if getenv("TMP") returns NULL?

File: util/env_windows.cc [hunk 16]
```diff
@@ -0,0 +1,920 @@
...
+  }
+
+  // Add to priority queue
+  queue_.push_back(BGItem());
+  queue_.back().function = function;
+  queue_.back().arg = arg;
+}
+
+void WindowsEnv::BGThread() {
+  while (started_bgthread_) {
```
Comment (ghemawat) [inline util/env_windows.cc]: This is accessing started_bgthread_ without holding a lock, and it is written by another thread.

The complication of termination should go away once you clean up the destructor above so it crashes without trying to shutdown this thread.

File: util/env_windows.cc [hunk 17]
```diff
@@ -0,0 +1,920 @@
...
+  t.detach();
+}
+
+}   // namespace
+
+static std::once_flag once;
+static Env* default_env;
+static void InitDefaultEnv() { default_env = new WindowsEnv; }
+
+char* FileLogger::AppendDateTo(char* first, char* last)
```
Comment (ghemawat) [inline util/env_windows.cc]: See my comment on the other AppendDateTo implementation.

File: util/env_windows_test.cc [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+}  // namespace leveldb
+
+int main(int argc, char** argv) {
+  // All tests currently run with the same read-only file limits.
+  leveldb::EnvWindowsTest::SetFileLimits(leveldb::kReadOnlyFileLimit,
+                                       leveldb::kMMapLimit);
+  return leveldb::test::RunAllTests();
+}
```
Comment (ghemawat) [inline util/env_windows_test.cc]: Add newline at end of file.

File: util/env_windows_test_helper.h [hunk 19]
```diff
@@ -0,0 +1,29 @@
...
+
+#ifndef STORAGE_LEVELDB_UTIL_ENV_WINDOWS_TEST_HELPER_H_
+#define STORAGE_LEVELDB_UTIL_ENV_WINDOWS_TEST_HELPER_H_
+
+namespace leveldb {
+
+class EnvWindowsTest;
+
+// A helper for the Windows Env to facilitate testing.
+class EnvWindowsTestHelper {
```
Comment (ghemawat) [inline util/env_windows_test_helper.h]: I am questioning if we really need these features on windows.  At least the mmap support should be added in a separate pull request after performance analysis.

Does windows have small open-file limits?  If not, the read-only-fd-limit can also be dropped, and this entire header file can go away.

File: .appveyor.yml [hunk 20]
```diff
@@ -0,0 +1,40 @@
+# Build matrix / environment variables are explained on:
```
[the remark this answers is not in the document]
Comment (Alexhuszagh) [reply .appveyor.yml]: The CMake build configuration is initially based off @pwnall's fork, so that would be ideal.

File: db/c_test.c [hunk 21]
```diff
@@ -168,10 +160,7 @@ int main(int argc, char** argv) {
   CheckCondition(leveldb_major_version() >= 1);
   CheckCondition(leveldb_minor_version() >= 1);
 
-  snprintf(dbname, sizeof(dbname),
-           "%s/leveldb_c_test-%d",
-           GetTempDir(),
-           ((int) geteuid()));
+  tmpnam(dbname);
```
[the remark this answers is not in the document]
Comment (Alexhuszagh) [reply db/c_test.c]: `mktemp()` would be better than tmpnam, but still susceptible to race conditions (although with a unique template, that would be unlikely). I'll edit this, if I can find anyway to use `tmpfile` I will, but I sincerely doubt it.

Is there a particular pattern you would prefer? We can get the temporary directory fairly easy using environment variables, and then add a leveldb-specific prefix for the filename.

File: db/corruption_test.cc [hunk 22]
```diff
@@ -16,6 +16,7 @@
 #include "db/filename.h"
 #include "db/log_format.h"
 #include "db/version_set.h"
+#include "util/env_windows_test_helper.h"
```
[the remark this answers is not in the document]
Comment (Alexhuszagh) [reply db/corruption_test.cc]: Sure. Let me look at integrating this. I can aim for that early next week.

File: db/corruption_test.cc [hunk 23]
```diff
@@ -370,5 +375,10 @@ TEST(CorruptionTest, UnrelatedKeys) {
 }  // namespace leveldb
 
 int main(int argc, char** argv) {
+#ifdef _WIN32
+  // Disable memory-mapped files for Windows, since they cannot be overwritten,
+  // preventing file corruption for unittests.
```
[the remark this answers is not in the document]
Comment (Alexhuszagh) [reply db/corruption_test.cc]: Great idea, I will make these changes globally so all the changes are in port and env files. Thanks for your feedback.

File: db/db_test.cc [hunk 24]
```diff
@@ -1605,7 +1605,9 @@ TEST(DBTest, DestroyEmptyDir) {
   std::vector<std::string> children;
   ASSERT_OK(env.GetChildren(dbname, &children));
   // The POSIX env does not filter out '.' and '..' special files.
+#ifndef OS_WIN
```
[the remark this answers is not in the document]
Comment (Alexhuszagh) [reply db/db_test.cc]: Is there a guaranteed order to the return? Also, we need to check to guarantee that `children` is not empty, or calling `back` is undefined-behavior. The following code should solve both:

```cpp
  // Remove the "." and ".." files that the Posix env returns.
  while (!children.empty() && (children.back() == "." || children.back() == ".."))
    children.pop_back();
```

File: db/version_set.cc [hunk 25]
```diff
@@ -920,6 +920,10 @@ Status VersionSet::Recover(bool *save_manifest) {
     return Status::Corruption("CURRENT file does not end with newline");
   }
   current.resize(current.size() - 1);
+  // Fix "\r\n" newlines on Windows
```
[the remark this answers is not in the document]
Comment (Alexhuszagh) [reply db/version_set.cc]: I believe the carriage returns are due to opening files in `text` mode. Any file that is not opened in binary mode may freely replace '\n' with "\r\n" on Windows. I can re-work the file modes to only open in binary mode, and remove this check.

File: port/port.h [hunk 26]
```diff
@@ -10,8 +10,8 @@
 // Include the appropriate platform specific file below.  If you are
 // porting to a new platform, see "port_example.h" for documentation
 // of what the new port_<platform>.h file must provide.
-#if defined(LEVELDB_PLATFORM_POSIX)
-#  include "port/port_posix.h"
+#if defined(LEVELDB_PLATFORM_POSIX) || defined(LEVELDB_PLATFORM_WINDOWS)
+#  include "port/port_cpp11.h"
```
[the remark this answers is not in the document]
Comment (Alexhuszagh) [reply port/port.h]: Good idea, thanks for the feedback. Will do.

File: port/port_config.h.in [hunk 27]
```diff
@@ -0,0 +1,34 @@
+// Copyright 2017 The LevelDB Authors. All rights reserved.
```
[the remark this answers is not in the document]
Comment (Alexhuszagh) [reply port/port_config.h.in]: See [pwnall/cmake/port/port_config.h.in](https://github.com/pwnall/leveldb/blob/cmake/port/port_config.h.in).

File: port/port_cpp11.cc [hunk 28]
```diff
@@ -0,0 +1,97 @@
...
+CondVar::CondVar(Mutex* mu) :
+    mu_(mu) {
+  assert(mu_);
+}
+
+CondVar::~CondVar() {
+}
+
+void CondVar::Wait() {
+  std::unique_lock<std::mutex> lock(mu_->mu_, std::defer_lock);
```
[the remark this answers is not in the document]
Comment (Alexhuszagh) [reply port/port_cpp11.cc]: Done, and thank you.

File: port/port_cpp11.cc [hunk 29]
```diff
@@ -0,0 +1,97 @@
...
+  cv_.wait(lock);
+}
```
[the remark this answers is not in the document]
Comment (Alexhuszagh) [reply port/port_cpp11.cc]: @ghemawat It does not, according to [33.4.4.3.1] in the C++17 standard.

```
unique_lock(mutex_type& m, defer_lock_t) noexcept;
6. Effects: Constructs an object of type unique_lock. 
7. Postconditions: pm == addressof(m) and owns == false.
```
Also, looking at how [Clang](https://github.com/llvm-mirror/libcxx/blob/master/include/__mutex_base#L150) implements [this](https://github.com/llvm-mirror/libcxx/blob/master/include/__mutex_base#L133), we should be safe. Clang sets `__owns_` to `false` when `defer_lock` is passed to the constructor, avoiding unlocking the mutex on object destruction.

File: port/port_cpp11.h [hunk 30]
```diff
@@ -0,0 +1,170 @@
...
+  Mutex();
+  ~Mutex();
+
+  void Lock();
+  void Unlock();
+  void AssertHeld() { }
+
+ private:
+  friend class CondVar;
+  // critical sections are more efficient than mutexes
```
[the remark this answers is not in the document]
Comment (Alexhuszagh) [reply port/port_cpp11.h]: @ghemawat Removed, and the other stale comment below. Nice catch.

File: port/port_cpp11.h [hunk 31]
```diff
@@ -0,0 +1,170 @@
...
+  ~CondVar();
+  void Wait();
+  void Signal();
+  void SignalAll();
+ private:
+  Mutex* mu_;
+  std::condition_variable cv_;
+};
+
+// Storage for a lock-free pointer
```
[the remark this answers is not in the document]
Comment (Alexhuszagh) [reply port/port_cpp11.h]: @ghemawat Sure, will do early next week.

File: util/env_windows_test_helper.h (hunk 19 above)
Comment (Alexhuszagh) [reply util/env_windows_test_helper.h]: @ghemawat I believe there is no practical limit, I can remove this.

File: .appveyor.yml (hunk 20 above)
[the remark this answers is not in the document]
Comment (ghemawat) [reply .appveyor.yml]: @pwnall can you take a look at the config stuff?  Though maybe later since I think there will be a resubmitted PR after making the changes I suggested.

File: db/c_test.c (hunk 21 above)
[the remark this answers is not in the document]
Comment (ghemawat) [reply db/c_test.c]: Anything similar to the older pattern which mentioned leveldb and the test name would be fine.

Yes, there will be race conditions with mktemp, I think it is fine for a test.

File: db/corruption_test.cc (hunk 23 above)
[the remark this answers is not in the document]
Comment (ghemawat) [reply db/corruption_test.cc]: This might not be necessary if mmap support is deferred.

File: db/db_test.cc (hunk 24 above)
[the remark this answers is not in the document]
Comment (ghemawat) [reply db/db_test.cc]: Your code snippet looks good.

File: port/port_cpp11.cc (hunk 29 above)
[the remark this answers is not in the document]
Comment (ghemawat) [reply port/port_cpp11.cc]: Thanks.

**Comment by pwnall** [comment]
@danielgindi Please see the answer at https://github.com/google/leveldb/pull/589#issuecomment-458431653

**Comment by pwnall** [comment]
We've landed the official Windows support, based on code from Chrome. Going to close this PR now. Sorry we couldn't get to it sooner!
